### PR TITLE
chore: add solidity verifier workflow

### PR DESCRIPTION
**Pull request by jonybur**
# Description

## Problem\*

Resolves 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: .github/workflows/test-integration.yml [hunk 1]
```diff
@@ -1,40 +1,70 @@
...
+    branches:
+      - jb-test-solidity-verifier
+
+# on:
+# workflow_dispatch:
+# pull_request:
+#   paths:
+#     - ./compiler/integration-tests/**
+# schedule:
+#   - cron: "0 2 * * *" # Run nightly at 2 AM UTC
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:14]: This needs to be activated again.

File: .github/workflows/test-integration.yml [hunk 2]
```diff
@@ -1,40 +1,70 @@
...
         with:
-          nix_path: nixpkgs=channel:nixos-23.05
+          nix_path: nixpkgs=channel:nixos-22.11
           github_access_token: ${{ secrets.GITHUB_TOKEN }}
 
-      - uses: cachix/cachix-action@v12
-        with:
-          name: barretenberg
-          authToken: "${{ secrets.CACHIX_AUTH_TOKEN }}"
-
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:37]: why is this removed?

File: .github/workflows/test-integration.yml [hunk 3]
```diff
@@ -52,53 +82,209 @@ jobs:
...
+
+      - name: Run Anvil
+        run: |
+          anvil > /dev/null 2>&1 &
+          sleep 10
+
+      - name: Configure git
+        run: |
+          git config user.name kevaundray
+          git config user.email kevtheappdev@gmail.com
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:232]:
```suggestion
```

File: .github/workflows/test-integration.yml [hunk 4]
```diff
@@ -52,53 +82,209 @@ jobs:
...
+          dv_result=$(forge create --rpc-url http://127.0.0.1:8545 --mnemonic "test test test test test test test test test test test junk" src/double_verify_proof.sol:UltraVerifier)
+          echo "$dv_result"
+          address=$(echo "$dv_result" | grep "Deployed to:" | awk '{print $3}')
+          echo "DV_CONTRACT_ADDRESS=$address" >> $GITHUB_ENV
+
+      - name: Replace addresses in the file
+        working-directory: compiler/integration-tests/test/integration/browser
+        run: |
+          sed -i 's|address: "MUL_CONTRACT_ADDRESS"|address: "'"$MUL_CONTRACT_ADDRESS"'"|g' compile_prove_verify.test.ts
+          sed -i 's|address: "DV_CONTRACT_ADDRESS"|address: "'"$DV_CONTRACT_ADDRESS"'"|g' compile_prove_verify.test.ts
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:271]: These should be written to a separate file rather than overwriting the test file. This currently makes it a pain to run tests multiple times locally

File: .github/workflows/test-integration.yml [hunk 5]
```diff
@@ -52,53 +82,209 @@ jobs:
...
+      # - name: Alert on nightly test failure
+      #   uses: JasonEtco/create-an-issue@v2
+      #   if: ${{ failure() && github.event_name == 'schedule' }}
+      #   env:
+      #     GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      #     WORKFLOW_NAME: ${{ github.workflow }}
+      #     WORKFLOW_URL: ${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}
+      #   with:
+      #     update_existing: true
+      #     filename: .github/NIGHTLY_TEST_FAILURE.md
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:290]: Reinstate alerting

File: compiler/integration-tests/test/integration/browser/compile_prove_verify.test.ts [hunk 6]
```diff
@@ -4,14 +4,19 @@ import newCompiler, {
...
-import { acvm, noirc } from "@noir-lang/noir_js";
 import { decompressSync as gunzip } from "fflate";
-
+import { ethers } from "ethers";
 import * as TOML from "smol-toml";
 
+const mnemonic = "test test test test test test test test test test test junk";
+const provider = new ethers.JsonRpcProvider("http://localhost:8545");
+const walletMnemonic = ethers.Wallet.fromPhrase(mnemonic);
+const wallet = walletMnemonic.connect(provider);
```
Comment (TomAFrench) [inline compiler/integration-tests/test/integration/browser/compile_prove_verify.test.ts:16]: We should be verifying the proof in a `eth_call` context which removes the need for a full wallet. Not a blocker.

File: .github/workflows/test-integration.yml [hunk 7]
```diff
@@ -52,53 +82,209 @@ jobs:
...
+    runs-on: ubuntu-latest
+    needs: [build-wasm, build-nargo, build-noirc]
+    env:
+      CACHED_PATH: /tmp/nix-cache
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ inputs.tag || env.GITHUB_REF }}
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:181]: This input doesn't exist.

**Comment by TomAFrench** [comment]
There's a bunch of changes which need to be reverted due to experiementing in CI. We can clean up the actual testing flow in a followup PR.

File: .github/workflows/test-integration.yml [hunk 8]
```diff
@@ -1,40 +1,70 @@
-name: test-integration
+name: Test Solidity Verifier
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:1]:
```suggestion
name: test-integration
```

Let's revert this this workflow isn't just for the solidity verifiers.

File: .github/workflows/test-integration.yml [hunk 9]
```diff
@@ -52,53 +82,205 @@ jobs:
...
+      # - name: Alert on nightly test failure
+      #   uses: JasonEtco/create-an-issue@v2
+      #   if: ${{ failure() && github.event_name == 'schedule' }}
+      #   env:
+      #     GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      #     WORKFLOW_NAME: ${{ github.workflow }}
+      #     WORKFLOW_URL: ${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}
+      #   with:
+      #     update_existing: true
+      #     filename: .github/NIGHTLY_TEST_FAILURE.md
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:286]:
```suggestion
      - name: Alert on nightly test failure
        uses: JasonEtco/create-an-issue@v2
        if: ${{ failure() && github.event_name == 'schedule' }}
        env:
            GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
            WORKFLOW_NAME: ${{ github.workflow }}
            WORKFLOW_URL: ${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}
         with:
             update_existing: true
             filename: .github/NIGHTLY_TEST_FAILURE.md
```

File: .github/workflows/test-integration.yml [hunk 10]
```diff
@@ -52,53 +82,205 @@ jobs:
...
-          WORKFLOW_URL: ${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}
-        with:
-          update_existing: true
-          filename: .github/NIGHTLY_TEST_FAILURE.md
+           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+           WORKFLOW_NAME: ${{ github.workflow }}
+           WORKFLOW_URL: ${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}
+         with:
+           update_existing: true
+           filename: .github/NIGHTLY_TEST_FAILURE.md
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:286]:
```suggestion
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          WORKFLOW_NAME: ${{ github.workflow }}
          WORKFLOW_URL: ${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}
        with:
          update_existing: true
          filename: .github/NIGHTLY_TEST_FAILURE.md
```

File: .github/workflows/test-integration.yml [hunk 11]
```diff
@@ -52,45 +82,197 @@ jobs:
...
+
       - name: Setup `integration-tests`
         run: |
           yarn workspace @noir-lang/source-resolver build
           yarn workspace @noir-lang/noir_js build
 
       - name: Run `integration-tests`
-        run: |
-          yarn test:integration
+        run: yarn test:integration
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:275]:
```suggestion
        run: |
            yarn test:integration
```

File: .github/workflows/test-integration.yml [hunk 12]
```diff
@@ -52,37 +82,190 @@ jobs:
...
+          path: ./compiler/wasm
+
+      - name: Download noirc package artifact
+        uses: actions/download-artifact@v3
+        with:
+          name: noirc_abi_wasm
+          path: ./tooling/noirc_abi_wasm
+
+      - name: Set nargo on PATH
+        working-directory: ./compiler/wasm/noir-script
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:208]:
```suggestion
```

File: compiler/integration-tests/test/integration/browser/compile_prove_verify.test.ts [hunk 13]
```diff
@@ -159,7 +178,23 @@ test_cases.forEach((testInfo) => {
...
+        expect(verified, "Proof fails verification in JS").to.be.true;
+
+        try {
+          let result;
+          if (testInfo.publicInputsLength === 0) {
+            result = await contract.verify(proof, []);
+          } else {
+            const publicInputs = proof.slice(0, testInfo.publicInputsLength);
+            const slicedProof = proof.slice(testInfo.publicInputsLength);
+            result = await contract.verify(slicedProof, publicInputs);
```
Comment (TomAFrench) [inline compiler/integration-tests/test/integration/browser/compile_prove_verify.test.ts:190]: We don't actually have a test for this codepath. Can you add a new circuit with a public input? We have one [here](https://github.com/noir-lang/noir/tree/master/compiler/integration-tests/test/circuits/main) which is suitable.

File: .github/workflows/test-integration.yml [hunk 14]
```diff
@@ -28,38 +67,191 @@ jobs:
...
+          dv_result=$(forge create --rpc-url http://127.0.0.1:8545 --mnemonic "test test test test test test test test test test test junk" src/double_verify_proof.sol:UltraVerifier)
+          echo "$dv_result"
+          address=$(echo "$dv_result" | grep "Deployed to:" | awk '{print $3}')
+          echo "DV_CONTRACT_ADDRESS=$address" >> $GITHUB_ENV
+
+      - name: Replace addresses in the file
+        working-directory: compiler/integration-tests/test
+        run: |
+          sed -i 's|MUL_CONTRACT_ADDRESS = ""|MUL_CONTRACT_ADDRESS = "'"$MUL_CONTRACT_ADDRESS"'"|g' environment.js
+          sed -i 's|DV_CONTRACT_ADDRESS = ""|DV_CONTRACT_ADDRESS = "'"$DV_CONTRACT_ADDRESS"'"|g' environment.js
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:253]: We could instead create a JSON file of the form:

```
{
  "1_mul": "0x00..00",
  "double_verify_proof": "0x00..00"
}
```

This avoids us writing to a persistent file as the JSON can just be overwritten as needed.

File: .github/workflows/test-integration.yml [hunk 15]
```diff
@@ -28,38 +67,191 @@ jobs:
...
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Checkout acvm sources
+        uses: actions/checkout@v3 # v3 is needed here otherwise this fails in local execution
+        with:
+          repository: noir-lang/acvm
+          path: acvm
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:175]:
```suggestion
```

File: .github/workflows/test-integration.yml [hunk 16]
```diff
@@ -28,38 +67,191 @@ jobs:
...
+        run: |
+          mul_result=$(forge create --rpc-url http://127.0.0.1:8545 --mnemonic "test test test test test test test test test test test junk" src/1_mul.sol:UltraVerifier)
+          echo "$mul_result"
+          address=$(echo "$mul_result" | grep "Deployed to:" | awk '{print $3}')
+          echo "MUL_CONTRACT_ADDRESS=$address" >> $GITHUB_ENV
+
+          dv_result=$(forge create --rpc-url http://127.0.0.1:8545 --mnemonic "test test test test test test test test test test test junk" src/double_verify_proof.sol:UltraVerifier)
+          echo "$dv_result"
+          address=$(echo "$dv_result" | grep "Deployed to:" | awk '{print $3}')
+          echo "DV_CONTRACT_ADDRESS=$address" >> $GITHUB_ENV
```
Comment (TomAFrench) [inline .github/workflows/test-integration.yml:247]: `forge create` supports a json output which would be simpler to read the address from.

File: .github/workflows/test-integration.yml (hunk 14 above)
Comment (jonybur) [reply .github/workflows/test-integration.yml:248]: I just followed the current pattern of having the `environment.js` exporting a const

File: .github/workflows/test-integration.yml (hunk 16 above)
Comment (jonybur) [reply .github/workflows/test-integration.yml:242]: Will replace this implementation with the json output of forge create

**Comment by jonybur** [comment]
Recursion test is failing but I think that's out of scope for this ticket

File: compiler/integration-tests/test/integration/browser/compile_prove_verify.test.ts [hunk 17]
```diff
@@ -159,7 +182,23 @@ test_cases.forEach((testInfo) => {
...
+        expect(verified, "Proof fails verification in JS").to.be.true;
+
+        try {
+          let result;
+          if (testInfo.publicInputsLength === 0) {
+            result = await contract.verify(proof, []);
+          } else {
+            const publicInputs = proof.slice(0, testInfo.publicInputsLength);
+            const slicedProof = proof.slice(testInfo.publicInputsLength);
+            result = await contract.verify(slicedProof, [publicInputs]);
```
Comment (TomAFrench) [inline compiler/integration-tests/test/integration/browser/compile_prove_verify.test.ts:194]: This doesn't look correct. What you're passing is an array with a single element which is a `Uint8Array` (which in this case is 32 bytes long). The verifier contract expects a `bytes32[]` so this works only if we have a single public input. If we used a circuit with two public inputs then this will start failing as you'll try writing 64 bytes into a `bytes32`.

We need a `Uint8Array[]` rather than a `[Uint8Array]`

**Review by TomAFrench — APPROVED** [review]
